honeybadgermpc/batch_reconstruction.py

```diff
@@ -4,51 +4,127 @@
+
+    # Pad final chunk with default value
+    res[-1] += [default] * (chunk_size - len(res[-1]))
+
+    return res
+
+
+def flatten_lists(lists):
+    """ Given a 2d list, return a flattened 1d list
+    e.g. [[1,2,3],[4,5,6],[7,8,9]] => [1,2,3,4,5,6,7,8,9]
+    """
+    return [elem for inner in lists for elem in inner]
+
+
+def transpose_lists(lists):
+    """ Given a 2d list, return the transpose of the list
+    e.g. [[1,2,3],[4,5,6],[7,8,9]] => [[1,4,7],[2,5,8],[3,6,9]]
+    """
+    rows = len(lists)
+    cols = len(lists[0])
+    return [[lists[j][i] for j in range(rows)] for i in range(cols)]
+
+
+async def fetch_one(awaitables):
+    """ Given a list of awaitables, run them concurrently and
+    return them in the order they complete
+
+    args:
+        awaitables: List of tasks to run concurrently
+
+    output:
+        Yields tuples of the form (idx, result) in the order that the tasks finish
+    """
+    mapping = {elem: idx for (idx, elem) in enumerate(awaitables)}
+    pending = set(awaitables)
     while len(pending) > 0:
-        done, pending = await asyncio.wait(pending,
-                                           return_when=asyncio.FIRST_COMPLETED)
+        done, pending = await asyncio.wait(pending, return_when=asyncio.FIRST_COMPLETED)
         for d in done:
-            yield (aws_to_idx[d], await d)
+            yield(mapping[d], await d)
+
+
+async def incremental_decode(receivers, encoder, decoder, robust_decoder, batch_size, t,
+                             n):
+    inc_decoder = IncrementalDecoder(encoder, decoder, robust_decoder,
+                                     degree=t, batch_size=batch_size,
+                                     max_errors=t)
+
+    async for idx, d in fetch_one(receivers):
+        inc_decoder.add(idx, d)
+        if inc_decoder.done():
+            result, _ = inc_decoder.get_results()
+            return result
+
+    return None
 
 
 def subscribe_recv(recv):
+    """ Given the recv method for this batch reconstruction,
+    create a background loop to put the received events into
+    the appropriate queue for the tag
+
+    Returns _task and subscribe, where _task is to be run in
+    the background to forward events to the associated queue,
+    and subscribe, which is used to register a new tag/queue pair
+    """
+    # Stores the queues for each subscribed tag
     tag_table = defaultdict(Queue)
     taken = set()  # Replace this with a bloom filter?
 
     async def _recv_loop():
         while True:
+            # Whenever we receive a share array, directly put it in the
+            # appropriate queue for that round
             j, (tag, o) = await recv()
             tag_table[tag].put_nowait((j, o))
 
     def subscribe(tag):
-        # take everything from the queue
-        # further things sent directly
+        # Ensure that this tag has not been subscribed to already
         assert tag not in taken
```

Comment:
NYC: This should not be an assert, this should raise an Exception - we disable asserts when running and this would break a lot of things if the messages went to the wrong queue.

Reply:
I don't think this is a massive priority right now, since the only tags used as of now are `"S"`, `"R1"`, and  `"R2"`. I agree on principle though.

Reply:
This code is actually reused in a lot of other places and not just batch reconstruction. We create one tag per run of AVSS, ACS for the offline phase and that is expected to run endlessly. Since it is a very small change, I'd suggest to take it.